source/common/network/connection_impl.cc

```diff
@@ -679,8 +679,15 @@ void ConnectionImpl::onWriteReady() {
       delayed_close_timer_->enableTimer(delayed_close_timeout_);
     }
     if (result.bytes_processed_ > 0) {
-      for (BytesSentCb& cb : bytes_sent_callbacks_) {
-        cb(result.bytes_processed_);
+      auto it = bytes_sent_callbacks_.begin();
+      while (it != bytes_sent_callbacks_.end()) {
+        if ((*it)(result.bytes_processed_)) {
+          // move to the next callback.
+          it++;
+        } else {
+          // remove the current callback.
+          it = bytes_sent_callbacks_.erase(it);
+        }
```

Comment:
This change itself looks risky, can we have this part as a separate PR to get reviewed?

Reply:
Agree. Created #14245.